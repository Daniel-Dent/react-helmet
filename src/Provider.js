import { Component } from 'react';
import PropTypes from 'prop-types';
import mapStateOnServer from './server';

export const providerShape = {
  helmet: PropTypes.func,
  helmetInstances: PropTypes.shape({
    get: PropTypes.func,
    add: PropTypes.func,
    remove: PropTypes.func,
  }),
};

const canUseDOM = typeof document !== 'undefined';

export default class Provider extends Component {
  static canUseDOM = canUseDOM;

  static propTypes = {
    context: PropTypes.shape({}),
    children: PropTypes.node.isRequired,
  };

  static defaultProps = {
    context: {},
  };

  static childContextTypes = providerShape;

  instances = [];

  getChildContext() {
    return {
      helmet: store => {
        // eslint-disable-next-line react/prop-types
        this.props.context.helmet = store;
      },
      helmetInstances: {
        get: () => this.instances,
        add: instance => {
          this.instances.push(instance);
        },
        remove: instance => {
          const index = this.instances.indexOf(instance);
          this.instances.splice(index, 1);
        },
      },
    };
  }

  constructor(props) {
    super(props);

    if (!Provider.canUseDOM) {
      props.context.helmet = mapStateOnServer({
        baseTag: [],
        bodyAttributes: {},
        encodeSpecialCharacters: true,
        htmlAttributes: {},
        linkTags: [],
        metaTags: [],
        noscriptTags: [],
        scriptTags: [],
        styleTags: [],
        title: '',
        titleAttributes: {},
      });
    }
  }

  render() {
    return this.props.children;
  }
}
