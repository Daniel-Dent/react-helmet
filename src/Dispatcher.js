import { Component } from 'react';
import shallowEqual from 'shallowequal';
import handleStateChangeOnClient from './client';
import mapStateOnServer from './server';
import { reducePropsToState } from './utils';
import Provider, { providerShape } from './Provider';

export default class Dispatcher extends Component {
  static contextTypes = providerShape;

  shouldComponentUpdate(nextProps) {
    return !shallowEqual(nextProps, this.props);
  }

  emitChange() {
    const { helmetInstances, setHelmet } = this.context;
    let serverState = null;
    const state = reducePropsToState(helmetInstances.get().map(instance => instance.props));
    if (Provider.canUseDOM) {
      handleStateChangeOnClient(state);
    } else if (mapStateOnServer) {
      serverState = mapStateOnServer(state);
    }
    setHelmet(serverState, state);
  }

  componentWillMount() {
    const { helmetInstances } = this.context;
    helmetInstances.add(this);
    this.emitChange();
  }

  componentDidUpdate() {
    this.emitChange();
  }

  componentWillUnmount() {
    const { helmetInstances } = this.context;
    helmetInstances.remove(this);
    this.emitChange();
  }

  render() {
    return null;
  }
}
