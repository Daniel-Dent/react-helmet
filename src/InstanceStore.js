import { Component } from 'react';
import shallowEqual from 'shallowequal';
import handleStateChangeOnClient from './client';
import mapStateOnServer from './server';
import { reducePropsToState } from './utils';
import Provider, { providerShape } from './Provider';

export default class InstanceStore extends Component {
  static contextTypes = providerShape;

  shouldComponentUpdate(nextProps) {
    return !shallowEqual(nextProps, this.props);
  }

  emitChange() {
    const { helmetInstances } = this.context;
    let state = reducePropsToState(helmetInstances.get().map(instance => instance.props));
    if (Provider.canUseDOM) {
      handleStateChangeOnClient(state);
    } else if (mapStateOnServer) {
      state = mapStateOnServer(state);
    }
    this.context.helmet(state);
  }

  componentWillMount() {
    const { helmetInstances } = this.context;
    helmetInstances.add(this);
    this.emitChange();
  }

  componentDidUpdate() {
    this.emitChange();
  }

  componentWillUnmount() {
    const { helmetInstances } = this.context;
    helmetInstances.remove(this);
    this.emitChange();
  }

  render() {
    return null;
  }
}
