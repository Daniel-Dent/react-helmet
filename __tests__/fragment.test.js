import React, { Fragment } from 'react';
import ReactDOM from 'react-dom';
import { Helmet } from '../src';
import Provider from '../src/Provider';

Helmet.defaultProps.defer = false;

const mount = document.getElementById('mount');

const render = node => {
  ReactDOM.render(<Provider>{node}</Provider>, mount);
};

describe('fragments', () => {
  it('parses Fragments', () => {
    render(
      <Helmet>
        <Fragment>
          <title>Hello</title>
        </Fragment>
      </Helmet>
    );

    expect(document.title).toMatchSnapshot();
  });

  it('parses nested Fragments', () => {
    render(
      <Helmet>
        <Fragment>
          <title>Foo</title>
          <Fragment>
            <title>Bar</title>
            <Fragment>
              <title>Baz</title>
            </Fragment>
          </Fragment>
        </Fragment>
      </Helmet>
    );

    expect(document.title).toMatchSnapshot();
  });
});
