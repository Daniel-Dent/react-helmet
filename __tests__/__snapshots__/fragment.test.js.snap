// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`misc parses Fragments 1`] = `"Hello"`;

exports[`misc parses nested Fragments 1`] = `"Baz"`;
